File: src/career_os/schemas/onboarding.py
"""Pydantic schemas for onboarding API."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

# Canonical step names — must match OnboardingState column names (without _at suffix).
# Order defines next_step computation in the service layer (STEP_ORDER in services/onboarding.py).
VALID_STEPS: list[str] = [
    "profile_started",
    "profile_completed",
    "demo_seeded",
    "welcome_completed",
    "tour_completed",
    "feedback_prompted",
    "completed",
]


class OnboardingStepUpdate(BaseModel):
    """Request body for PATCH /api/onboarding/status (D-05).

    Marks a single step complete. The server sets the timestamp (D-05).
    Re-patching the same step is a no-op (D-06) — handled in service layer.
    Step ordering is NOT enforced (D-07).
    """

    step: str = Field(
        ...,
        description=(f"Step name to mark complete. Valid values: {', '.join(VALID_STEPS)}"),
    )
    via: Literal["cli", "web"] = Field(
        ...,
        description="Surface that completed this step ('cli' or 'web').",
    )


class OnboardingStatusResponse(BaseModel):
    """Response schema for GET and PATCH /api/onboarding/status (D-04).

    Computed fields (next_step, is_complete, progress_pct) are populated
    by the service layer — callers do not need to derive these.
    """

    profile_id: int
    current_step: str | None = None

    # Per-step completion timestamps (D-01) — None means not yet completed
    profile_started_at: datetime | None = None
    profile_completed_at: datetime | None = None
    demo_seeded_at: datetime | None = None
    welcome_completed_at: datetime | None = None
    tour_completed_at: datetime | None = None
    feedback_prompted_at: datetime | None = None
    completed_at: datetime | None = None

    # Per-step source surface (D-02) — None means not completed
    profile_started_via: str | None = None
    profile_completed_via: str | None = None
    demo_seeded_via: str | None = None
    welcome_completed_via: str | None = None
    tour_completed_via: str | None = None
    feedback_prompted_via: str | None = None
    completed_via: str | None = None

    # Computed fields populated by service layer (D-04)
    next_step: str | None = None
    is_complete: bool = False
    progress_pct: int = 0

    # Audit timestamps
    created_at: datetime | None = None
    updated_at: datetime | None = None

    model_config = {"from_attributes": True}

File: frontend/src/api/onboarding.ts
/**
 * API client functions for onboarding status.
 */

/** Default profile ID (single-user self-hosted model). */
export const DEFAULT_PROFILE_ID = 1;

export interface OnboardingStatus {
  profile_id: number;
  current_step: string | null;
  next_step: string | null;
  is_complete: boolean;
  progress_pct: number;
  profile_started_at: string | null;
  profile_completed_at: string | null;
  demo_seeded_at: string | null;
  welcome_completed_at: string | null;
  tour_completed_at: string | null;
  feedback_prompted_at: string | null;
  completed_at: string | null;
  profile_started_via: string | null;
  profile_completed_via: string | null;
  demo_seeded_via: string | null;
  welcome_completed_via: string | null;
  tour_completed_via: string | null;
  feedback_prompted_via: string | null;
  completed_via: string | null;
  created_at: string | null;
  updated_at: string | null;
}

const API_BASE = "/api/onboarding";

/** Fetch onboarding status for a profile. */
export async function fetchOnboardingStatus(
  profileId: number,
): Promise<OnboardingStatus> {
  const res = await fetch(`${API_BASE}/status?profile_id=${profileId}`);
  if (!res.ok) {
    throw new Error(`Failed to fetch onboarding status: ${res.status}`);
  }
  return res.json() as Promise<OnboardingStatus>;
}

/** Mark an onboarding step complete. */
export async function patchOnboardingStep(
  profileId: number,
  step: string,
): Promise<OnboardingStatus> {
  const res = await fetch(`${API_BASE}/status?profile_id=${profileId}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ step, via: "web" }),
  });
  if (!res.ok) {
    throw new Error(`Failed to update onboarding step: ${res.status}`);
  }
  return res.json() as Promise<OnboardingStatus>;
}

/** Reset the onboarding flow (keeps profile data, restarts welcome/tour). */
export async function resetOnboarding(
  profileId: number,
): Promise<OnboardingStatus> {
  const res = await fetch(`${API_BASE}/reset?profile_id=${profileId}`, {
    method: "POST",
  });
  if (!res.ok) {
    throw new Error(`Failed to reset onboarding: ${res.status}`);
  }
  return res.json() as Promise<OnboardingStatus>;
}
